docs/developer-guide/custom-widgets/universal-widgets.md

```diff
@@ -0,0 +1,339 @@
+        if (oldProps.className) el.classList.remove(oldProps.className);
+        if (props.className) el.classList.add(props.className);
+      }
+    }
+    ...
+  }
+}
+```
+
+### CSS Variables
+
+Define variables in the widget's CSS stylesheet
+
+```css
+.my-awesome-widget {
+  background-color: var(--primary-color, rebeccapurple);
+}
+```
+
+Override in user's application
+
+```css
+.my-awesome-widget {
+    --primary-color: chartreuse;
+}
+```
+
+## Example: Layer List Widget with Preact
+
+Below is a comprehensive example demonstrating a layer list widget implemented using Preact for dynamic UI rendering:
+
+```tsx
+import {
+  _deepEqual as deepEqual,
+  _applyStyles as applyStyles,
+  _removeStyles as removeStyles
+} from '@deck.gl/core'
+import type {
+  Deck, Viewport, Widget, WidgetPlacement, Layer
+} from '@deck.gl/core'
+import {render} from 'preact';
+
+interface LayerListWidgetProps {
+  id?: string;
+  /**
+   * Widget positioning within the view. Default: 'top-left'.
+   */
+  placement?: WidgetPlacement;
+  /**
+   * View to attach to and interact with. Required when using multiple views. Default: null
+   */
+  viewId?: string | null;
+  /**
+   * CSS inline style overrides.
+   */
+  style?: Partial<CSSStyleDeclaration>;
+  /**
+   * Additional CSS class.
+   */
+  className?: string;
+}
+
+class LayerListWidget implements Widget {
+  id = 'layer-list-widget';
+  props: ZoomWidgetProps;
+  placement: WidgetPlacement = 'top-left';
+  viewId?: string | null = null;
+  viewports: {[id: string]: Viewport} = {};
+  layers: Layer[] = [];
+  deck?: Deck<any>;
+  element?: HTMLDivElement;
+
+  constructor(props: AwesomeWidgetProps) {
+    this.id = props.id || 'layer-list-widget';
+    this.placement = props.placement || 'top-left';
+    this.viewId = props.viewId || null;
+
+    this.props = { 
+      ...props,
+      style: props.style || {}
+    }
+  }
+
+  onAdd({deck}: {deck: Deck<any>}): HTMLDivElement {
+    const {style, className} = this.props;
+    const element = document.createElement('div');
+    element.classList.add('deck-widget', 'deck-widget-zoom');
+    if (className) element.classList.add(className);
+    applyStyles(element, style);
+    this.deck = deck;
+    this.element = element;
+    this.update();
+    return element;
+  }
+
+  onRedraw({layers}: {layers: Layer[]}) {
+    this.layers = layers;
+    this.update();
+  }
+
+  onViewportChange(viewport) {
+    this.viewports[viewport.id] = viewport
+  }
+
+  update() {
+    const element = this.element;
+    if (!element) {
+      return;
+    }
+    let layers = this.layers
+    if (this.deck?.props.layerFilter) {
+      const ui = (
+        {viewports.values().map(viewport => (
+          <div>
+            {viewport.id}
+            <ul>
+              {layers.filter(layer => (
+                this.deck?.props.layerFilter({layer, viewport})
+              )).map((layer) => {
+                <li key={layer.id}>{layer.id}</li>
```

Comment:
Would it be much more code to show how control the layers in some way? For example, have a checkbox next to each layer to toggle visibility in that view?

Reply:
I can try it.. but controlling anything the user sets starts to get into a territory where user state and internal state can get out of sync. At least, there's nothing preventing user-space from just overriding any changes a widget makes to a layer. And there's some risk of introducing a race condition.

I think what this would look like with our current API is something like:

```
this.deck?.setProps({ layers: layers.map(layer => layer.clone({ visible: isVisible }) })
```

This could cause events to fire in the app that in-turn call `setProps`. Is there a way to draw the layers without alerting the user? Perhaps:
```
this.deck?._drawLayers('layer-list-widget', { layers: ... })
```